x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/utils/ToXContentCompressor.java

```diff
@@ -0,0 +1,78 @@
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.Map;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Collection of helper methods. Similar to CompressedXContent, but this utilizes GZIP for parity with the native compression
+ */
+public final class ToXContentCompressor {
+    private static final int BUFFER_SIZE = 4096;
+
+    private ToXContentCompressor() {}
+
+    public static <T extends ToXContentObject> String deflate(T objectToCompress) throws IOException {
+        BytesReference reference = XContentHelper.toXContent(objectToCompress, XContentType.JSON, false);
+        return deflate(reference);
+    }
+
+    public static <T> T inflate(String compressedString,
+                                CheckedFunction<XContentParser, T, IOException> parserFunction,
+                                NamedXContentRegistry xContentRegistry) throws IOException {
+        try(XContentParser parser = XContentHelper.createParser(xContentRegistry,
+            LoggingDeprecationHandler.INSTANCE,
+            inflate(compressedString),
+            XContentType.JSON)) {
+            return parserFunction.apply(parser);
+        }
+    }
+
+    public static Map<String, Object> inflateToMap(String compressedString) throws IOException {
+        // Don't need the xcontent registry as we are now deflating named objects.
+        try(XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY,
+            LoggingDeprecationHandler.INSTANCE,
+            inflate(compressedString),
+            XContentType.JSON)) {
+            return parser.mapOrdered();
+        }
+    }
+
+    static BytesReference inflate(String compressedString) throws IOException {
+        byte[] compressedBytes = Base64.getDecoder().decode(compressedString.getBytes(StandardCharsets.UTF_8));
+        InputStream inflateStream = new GZIPInputStream(new BytesArray(compressedBytes).streamInput(), BUFFER_SIZE);
+        return Streams.readFully(inflateStream);
+    }
+
+    //Public for testing (for now)
+    public static String deflate(BytesReference reference) throws IOException {
+        BytesStreamOutput out = new BytesStreamOutput();
+        try (OutputStream compressedOutput = new GZIPOutputStream(out, BUFFER_SIZE)) {
+            reference.writeTo(compressedOutput);
+        }
+        return new String(Base64.getEncoder().encode(BytesReference.toBytes(out.bytes())), StandardCharsets.UTF_8);
```

Comment:
I would add a header/magic, e.g. just 1 byte. This will ensure future proof if you at some point in future want to change the format, which can be different compression, different baseXX encoding or something else. If you think base64 will never be changed the header can be in out and e.g. start with `01`, otherwise you need a printable/json-friendly version. 

The other thing is a truncation guard. I think neither gzipoutputstream nor base64 is safe regarding truncation. So the header should store the length, I would use a variable length encoding for that.

In summary:

{COMPRESSION_HEADER_BYTE}{VINT_ENCODED_LENGTH}{COMPRESSED_DEFINITION}

inflate of course needs the counterpart, compression header is just a check for now, in future this might branch to different implementations. The encoded length however should be checked, so you do not even have to try to parse it if you find a mismatch.

Additionally a the length should have a safe guard, with other words: define a maximum allowed length.